Introduce CustomSlurmSettings in SlurmSettings

jdeamicis opened the pull request:
### Description of changes
* Introduce `CustomSlurmSettings` and `CustomSlurmSettingsIncludeFile` under `SlurmSettings`.
* Add validation for `CustomSlurmSettings` at bulk slurm.conf level.
* Validate that `CustomSlurmSettingsIncludeFile` is a valid URL and that it is not used together with `CustomSlurmSettings`.
* Introduce context in `SLURM_SETTINGS_DENY_LIST` dictionary to differentiate groups of denied parameters based on when they are used.
* Reformat `SLURM_SETTINGS_DENY_LIST` for better readability of the structure of the dictionary.
* Refactor `SlurmCustomSettings*` objects to `CustomSlurmSettings*`.

### Tests
* Add unit test cases for the `CustomSlurmSettings` at `SlurmConf` level.
* Add unit tests for `CustomSlurmSettingsIncludeFileOnlyValidator`.
* Successfully performed manual test with new configuration parameter formatted this way:
  ```
  Scheduling:
  Scheduler: slurm
  SlurmSettings:
    CustomSlurmSettings:
      - Param1: Value1
      - Param2: Value2
      - Nodename: node-bla-[001-100]
        CPUs: 16
  ```

### References
* https://github.com/aws/aws-parallelcluster/pull/5079

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] ~Check if documentation is impacted by this change.~   Documentation coming at a second time.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Introduce CustomSlurmSettings in SlurmSettings

Add CustomSlurmSettings parameter to SlurmSettings schema.

Add custom_slurm_settings object to SlurmSettings class.

Add validator for CustomSlurmSettings at bulk slurm.conf level

Introduce context in SLURM_SETTINGS_DENY_LIST dictionary to differentiate
groups of denied parameters based on when they are used.

Reformat SLURM_SETTINGS_DENY_LIST for better readability of the
structure of the dictionary.

Refactor SlurmCustomSettings* objects to CustomSlurmSettings*.

Add unit test cases for the CustomSlurmSettings at SlurmConf level.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Introduce CustomSlurmSettingsIncludeFile in SlurmSettings

Validate CustomSlurmSettingsIncludeFile as an URL. Also, check that it
is not used together with CustomSlurmSettings under SlurmSettings.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -10,40 +10,86 @@
+            "slurmctldlogfile",
+            "slurmdlogfile",
+            "taskplugin",
+            "jobcomppass",
```
[Nitpick] Keeping these values in alphabetical order will make it easier to maintain the list

jdeamicis replied:
Addressing in the next commit.

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -10,40 +10,86 @@
     This enum defines the scope where the custom settings are defined.
     """
 
+    SLURM_CONF = "SlurmConf"
```
Q: having here a more fine grained list o levels, like `SlurmConf.Global` and `SlurmConf.Accounting` would make the messaging more easy to understand or it would simply add some noise ?

jdeamicis replied:
The Context entity I added is more of an internal representation of when parameters are used more than something immediately recognizable by the end users (such as the Level you had already introduced).

I would say it would not be too useful to add it to the message.

Unless we want to make it explicit that some parameters are not allowed when our Slurm accounting integration is enabled. But for that we would need to re-write the messaging (possibly using a separate validator function).

jdeamicis replied:
Let's think about this.

NSsirena replied:
Ok. I'm fine with this.
We will change this behavior if/when there will be a clear need for it.

jdeamicis committed:
Add CustomSlurmNodeNamesValidator

Check that nodes passed via SlurmSettings/CustomSlurmSettings do not
contain either the `-st-` or `-dy-` patterns in their names.

Add unit tests for the new validator.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add validators for custom Slurm settings to test_all_validators

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -117,6 +117,34 @@ def _validate(self):
+        if bad_nodelists:
+            nodelists = ", ".join(sorted(bad_nodelists))
+            self._add_failure(
+                f"The following custom Slurm node names are not valid: {nodelists}",
```
Should we rephrase this messaging with a more user-friendly message like "Use of substrings `-st-` and `-dy-` is reserved for managed nodes. Following ...."

jdeamicis replied:
Addressing.

jdeamicis replied:
Done.

NSsirena commented on cli/src/pcluster/validators/slurm_settings_validator.py:
```diff
@@ -114,3 +115,20 @@ def _validate(self):
+        if custom_settings and include_file_url:
+            self._add_failure(
+                "CustomSlurmsettings and CustomSlurmSettingsIncludeFile cannot be used together "
+                "under SlurmSettings.",
```
I know they can only be specified in this section, however I'd remove the last part of the sentence `under SlurmSettings` ...since they cannot be used together anywhere and this message can be misleading.

jdeamicis replied:
Actually they can use the `CustomSlurmSettings` at Queue and ComputeResource level while still using `CustomSlurmSettingsIncludeFile`, since they are in two different parts of the Slurm configuration managed by ParallelCluster.

NSsirena commented on cli/tests/pcluster/validators/test_all_validators.py:
Why are we removing `scheduler_plugin` tests ?

They should be removed as part of another story/task ..they are unrelated to this feature.

jdeamicis replied:
Moved to #5111
